### PR TITLE
BTAT-7948 Added new route controller model

### PR DESCRIPTION
**Pull request by JBlacklawsTen10**
# Pull Request Checklist
* [X] Branch is rebased with master
* [X] Added Unit Tests for changed functionality
* [X] Ran `sbt clean compile coverage test it:test coverageReport`, all tests pass and coverage meets minimum
* [X] Checked libraries are up to date with `sbt validate`. (If applicable) library updates are done in a separate, single commit
* [X] Ran Acceptance tests as per the story. Query in group slack if unsure of test packs to run.
* [X] Ran Performance tests as per the story. Query in group slack if unsure of test packs to run.
* [ ] (If applicable) raised app-config-* PRs
* [ ] (If applicable) added Integration Test

### CODE REVIEW

File: test/uk/gov/hmrc/vatsubscription/models/updateVatSubscription/request/UpdateVatSubscriptionSpec.scala [hunk 1]
```diff
@@ -17,25 +17,44 @@
...
 import uk.gov.hmrc.vatsubscription.assets.TestUtil
-import play.api.libs.json.Json
+import play.api.libs.json.{JsValue, Json}
+import uk.gov.hmrc.vatsubscription.helpers.{DeclarationTestConstants, DeregistrationInfoTestConstants}
 import uk.gov.hmrc.vatsubscription.helpers.UpdateVatSubscriptionTestConstants._
 import uk.gov.hmrc.vatsubscription.models.{MTDfBExempt, MTDfBMandated}
 
 class UpdateVatSubscriptionSpec extends TestUtil {
 
+  val updateVatSubscriptionLatestDESApi1365JsonMax: JsValue = Json.obj(
```
Comment (lewishay) [inline test/uk/gov/hmrc/vatsubscription/models/updateVatSubscription/request/UpdateVatSubscriptionSpec.scala:27]: As mentioned in an earlier comment, the writes in `UpdateVatSubscription` is not correct for the `commsPreference` field. If after fixing the writes you update this JSON here ("JsonMax" here is referring to every possible field being populated) to include the field `commsPreference`, then the test below will then be properly checking that the field is being written correctly.

File: test/uk/gov/hmrc/vatsubscription/models/updateVatSubscription/request/UpdateVatSubscriptionSpec.scala [hunk 2]
```diff
@@ -17,25 +17,44 @@
...
+
   "UpdateVatSubscription" when {
 
     "serializing to JSON" should {
 
       "for the latest DES API1365 writes" should {
 
         "Output the correct JSON for UpdateVatSubscriptionModelMax" in {
-          UpdateVatSubscription.DESApi1365Writes.writes(updateVatSubscriptionModelMax) shouldBe
+          UpdateVatSubscription.DESApi1365Writes(mockAppConfig).writes(updateVatSubscriptionModelMax) shouldBe
```
Comment (lewishay) [inline test/uk/gov/hmrc/vatsubscription/models/updateVatSubscription/request/UpdateVatSubscriptionSpec.scala:52]: Similar to above, once the JSON is updated then the model here `updateVatSubscriptionModelMax` will also need to be updated so that `ContactPreference` is not `None`

File: test/uk/gov/hmrc/vatsubscription/service/UpdateContactPreferenceServiceSpec.scala [hunk 3]
```diff
@@ -0,0 +1,123 @@
...
+ * limitations under the License.
+ */
+
+package uk.gov.hmrc.vatsubscription.service
+
+import uk.gov.hmrc.vatsubscription.assets.TestUtil
+import uk.gov.hmrc.vatsubscription.connectors.mocks.MockUpdateVatSubscriptionConnector
+import uk.gov.hmrc.vatsubscription.helpers.BaseTestConstants.testUser
+import uk.gov.hmrc.vatsubscription.httpparsers.UpdateVatSubscriptionHttpParser.UpdateVatSubscriptionResponse
+import uk.gov.hmrc.vatsubscription.models.{CommsPreference, DigitalPreference, PaperPreference}
```
Comment (lewishay) [inline test/uk/gov/hmrc/vatsubscription/service/UpdateContactPreferenceServiceSpec.scala:23]: Unused imports in this file

File: app/uk/gov/hmrc/vatsubscription/models/updateVatSubscription/request/RequestedChanges.scala [hunk 4]
```diff
@@ -17,6 +17,9 @@
 package uk.gov.hmrc.vatsubscription.models.updateVatSubscription.request
 
 import play.api.libs.json._
+import uk.gov.hmrc.vatsubscription.config.AppConfig
+import uk.gov.hmrc.vatsubscription.config.featureSwitch.{Api1365Latest, Api1365PreRelease}
+import uk.gov.hmrc.vatsubscription.models.CustomerDetails.jsonObjNoNulls
```
Comment (lewishay) [inline app/uk/gov/hmrc/vatsubscription/models/updateVatSubscription/request/RequestedChanges.scala:22]: Unused import

File: app/uk/gov/hmrc/vatsubscription/models/PendingChanges.scala [hunk 5]
```diff
@@ -35,8 +35,7 @@ object PendingChanges {
   private val bankDetailsPath =  __ \ "bankDetails"
   private val returnPeriodPath = __ \ "returnPeriod"
   private val mandationStatusDesPath = __ \ "mandationStatus" \ "mandationStatus"
-  private val mandationStatusWritesPath = __ \ "mandationStatus"
-  private val commsPreferencePath = __ \ "commsPreference"
+  private val commsPreferenceDesPath = __ \ "commsPreference" \ "commsPreference"
```
Comment (lewishay) [inline app/uk/gov/hmrc/vatsubscription/models/PendingChanges.scala:38]: I'm a little worried these changes are showing up as they were the changes I made in the bug fix which are now on master. I'm not sure the rebase has been done properly

File: app/uk/gov/hmrc/vatsubscription/services/UpdateContactPreferenceService.scala [hunk 6]
```diff
@@ -0,0 +1,61 @@
...
+
+    val agentOrCapacitor: Option[AgentOrCapacitor] = user.arn.map(AgentOrCapacitor(_, agentContactDetails))
+
+    UpdateVatSubscription(
+      controlInformation = ControlInformation(welshIndicator),
+      requestedChanges = ChangeCommsPreference,
+      updatedPPOB = None,
+      updatedReturnPeriod = None,
+      updateDeregistrationInfo = None,
+      declaration = Declaration(agentOrCapacitor, Signing()),
```
[the remark this answers is not in the document]
Comment (lewishay) [reply app/uk/gov/hmrc/vatsubscription/services/UpdateContactPreferenceService.scala:54]: Still not resolved

File: app/uk/gov/hmrc/vatsubscription/services/UpdateContactPreferenceService.scala [hunk 7]
```diff
@@ -0,0 +1,61 @@
...
+  def updateContactPreference(updatedContactPreference: CommsPreferencePost, welshIndicator: Boolean)
+                             (implicit user: User[_], hc: HeaderCarrier, ec: ExecutionContext): Future[UpdateVatSubscriptionResponse] = {
+    val subscriptionModel = constructContactPreferenceModel(updatedContactPreference, welshIndicator)
+    Logger.debug(s"[UpdateContactPreferenceService][updateContactPreference]: updating contact preference for user with vrn - ${user.vrn}")
+    updateVatSubscriptionConnector.updateVatSubscription(user, subscriptionModel, hc)
+  }
+
+  def constructContactPreferenceModel(updatedContactPreference: CommsPreferencePost, welshIndicator: Boolean)
+                                     (implicit user: User[_]): UpdateVatSubscription = {
+    val agentContactDetails: Option[ContactDetails] =
```
[the remark this answers is not in the document]
Comment (lewishay) [reply app/uk/gov/hmrc/vatsubscription/services/UpdateContactPreferenceService.scala:43]: Still not resolved

File: conf/application.conf [hunk 8]
```diff
@@ -171,7 +171,7 @@ microservice {
 
 feature-switch {
     Api1363Version = "Latest"
-    Api1365Version = "Latest"
+    Api1365Version = "Pre-release"
```
[the remark this answers is not in the document]
Comment (lewishay) [reply conf/application.conf:174]: Still not resolved

File: test/uk/gov/hmrc/vatsubscription/controllers/UpdateContactPreferenceControllerSpec.scala [hunk 9]
```diff
@@ -0,0 +1,139 @@
...
+            mockUpdateContactPreference(CommsPreferencePost(PaperPreference, None))(Future.successful(Left(updateErrorResponse)))
+            status(res) shouldBe INTERNAL_SERVER_ERROR
+          }
+
+          "return the expected error model" in {
+            jsonBodyOf(res) shouldBe Json.toJson(updateErrorResponse)
+          }
+        }
+
+        "a valid return period is supplied but an error is returned instead of a welshIndicator" should {
```
[the remark this answers is not in the document]
Comment (lewishay) [reply test/uk/gov/hmrc/vatsubscription/controllers/UpdateContactPreferenceControllerSpec.scala:122]: Still not resolved

File: app/uk/gov/hmrc/vatsubscription/controllers/UpdateContactPreferenceController.scala [hunk 10]
```diff
@@ -0,0 +1,53 @@
...
+class UpdateContactPreferenceController @Inject()(VatAuthorised: VatAuthorised,
+                                                  updatedContactPreferenceService: UpdateContactPreferenceService,
+                                                  vatCustomerDetailsRetrievalService: VatCustomerDetailsRetrievalService,
+                                                  cc: ControllerComponents)
+                                                 (implicit ec: ExecutionContext) extends BackendController(cc)
+                                                  with MicroserviceBaseController {
+
+  def updateContactPreference(vrn: String): Action[AnyContent] = VatAuthorised.async(vrn) {
+    implicit user =>
+      parseJsonBody[CommsPreferencePost] match {
```
Comment (miranda-hawkes) [inline app/uk/gov/hmrc/vatsubscription/controllers/UpdateContactPreferenceController.scala:39]: I just tested this with postman with the following body (the valid body works fine!):
```
{
  "commsPreference": "gdggdf"
}
```
it returned a 500 and the following error body response:
```
{
    "statusCode": 500,
    "message": "gdggdf (of class java.lang.String)"
}
```
It looks like when the JSON is invalid, `parseJsonBody` isn't caught by the Left(_), it throws a `scala.MatchError` exception which isn't being caught

File: app/uk/gov/hmrc/vatsubscription/models/updateVatSubscription/request/RequestedChanges.scala [hunk 11]
```diff
@@ -46,6 +50,9 @@ object RequestedChanges {
       "correspDetails" -> model.correspDetails,
       "organisationDetails" -> model.organisationDetails,
       "mandationStatus" -> model.mandationStatus
-    )
+    ) ++ (appConfig.features.api1365Version() match {
+      case Api1365Latest => Json.obj("commsPreference" -> model.commsPreference)
```
Comment (miranda-hawkes) [inline app/uk/gov/hmrc/vatsubscription/models/updateVatSubscription/request/RequestedChanges.scala:54]: Not really part of this PR but there isn't a way to toggle the API version via manage VAT - do we want a small task to add this in? You can only choose 'Latest' for both 1363 and 1365 on the feature switch page

File: it/uk/gov/hmrc/vatsubscription/controllers/UpdateContactPreferenceControllerISpec.scala [hunk 12]
```diff
@@ -0,0 +1,87 @@
+/*
+ * Copyright 2018 HM Revenue & Customs
```
Comment (miranda-hawkes) [inline it/uk/gov/hmrc/vatsubscription/controllers/UpdateContactPreferenceControllerISpec.scala:2]: Header has wrong year

File: test/uk/gov/hmrc/vatsubscription/connectors/UpdateVatSubscriptionConnectorSpec.scala [hunk 13]
```diff
@@ -64,7 +65,6 @@ class UpdateVatSubscriptionConnectorSpec extends TestUtil with MockHttpClient {
         val result: Future[UpdateVatSubscriptionResponse] = connector.updateVatSubscription(testUser, requestModel, hc)
 
         await(result) shouldBe Right(SuccessModel("12345"))
-        connector.writes shouldBe UpdateVatSubscription.DESApi1365Writes
```
Comment (miranda-hawkes) [inline test/uk/gov/hmrc/vatsubscription/connectors/UpdateVatSubscriptionConnectorSpec.scala:67]: Any reason this was removed?

File: test/uk/gov/hmrc/vatsubscription/controllers/UpdateContactPreferenceControllerSpec.scala [hunk 14]
```diff
@@ -0,0 +1,139 @@
...
+          val res: Result = await(TestCommsPreference.updateContactPreference(testVatNumber)(digitalPref))
+
+          status(res) shouldBe OK
+          jsonBodyOf(res) shouldBe Json.toJson(updateSuccessResponse)
+        }
+      }
+
+      "return an error response" when {
+
+        "no json body is supplied for the PUT" should {
```
Comment (miranda-hawkes) [inline test/uk/gov/hmrc/vatsubscription/controllers/UpdateContactPreferenceControllerSpec.scala:91]: As mentioned above, need a test for JSON being invalid/comms preference not being digital or paper (once fixed) - manually testing returns part of an exception in the response message

File: app/uk/gov/hmrc/vatsubscription/models/updateVatSubscription/request/RequestedChanges.scala (hunk 11 above)
Comment (lewishay) [reply app/uk/gov/hmrc/vatsubscription/models/updateVatSubscription/request/RequestedChanges.scala:54]: If you see value in it then yeah I'm happy for a story to be raised. When I added the feature I wondered if we would ever need to toggle that in our test environments, but as there was no immediate need to do so I left it.

File: test/uk/gov/hmrc/vatsubscription/connectors/UpdateVatSubscriptionConnectorSpec.scala (hunk 13 above)
Comment (lewishay) [reply test/uk/gov/hmrc/vatsubscription/connectors/UpdateVatSubscriptionConnectorSpec.scala:67]: It fails due to the fact that the Writes now takes an `appConfig` parameter and when the two anonymous functions are compared for equality they have two different references. Jack and I couldn't find a way to rewrite this, and ultimately after looking at exactly what it was testing it didn't seem like it was an important test (bit confused why it was checking the writes when there is only 1 writes in `UpdateVatSubscription`?)

If you think this is still a valuable test though we could try to explore other ways to rewrite it.
